pkg/core/stateroot/module.go

```diff
@@ -165,3 +169,30 @@ func (s *Module) verifyWitness(r *state.MPTRoot) error {
 	s.mtx.Unlock()
 	return s.bc.VerifyWitness(h, r, &r.Witness[0], maxVerificationGAS)
 }
+
+// Traverse traverses local MPT nodes starting from the specified root down to its
+// children calling f for each serialised node until stop condition is satisfied.
+func (s *Module) Traverse(root util.Uint256, f func(nodeBytes []byte), stop func(node []byte) bool) error {
+	tr := mpt.NewTrie(mpt.NewHashNode(root), false, storage.NewMemCachedStore(s.Store))
+	return tr.Traverse(f, stop)
+}
+
+// RestoreMPTNode tries to replace HashNode specified by the path to its "unhashed"
+// counterpart. An error is returned if the path doesn't point to a missing HashNode
+// or provided counterpart has invalid hash.
+func (s *Module) RestoreMPTNode(path []byte, node mpt.Node) error {
+	err := s.mpt.RestoreHashNode(path, node)
+	if err != nil {
+		return err
+	}
+	// TODO: either 1) put node into storage right now
+	// or 2) use modified `Flush` after all nodes are processed
+	// currently option 1) is implemented
+	_ = s.Store.Put(append([]byte{byte(storage.DataMPT)}, node.Hash().BytesBE()...), node.Bytes())
```

Comment:
The last 4 bytes must contain reference counter (if it is enabled of course). For example, `Flush` can just take last 4 bytes and interpret it as a reference counter. I think checking if node is already in store and appending 4 zeros otherwise will do the job.

Reply:
Fixed. I also moved this part of code from `(s *Module) RestoreMPTNode` to `(t *Trie) RestoreHashNode`, because we use MPT key and this logic is more related to the MPT.

We definitely need a test for it, I'll write it.